Adds a docker image for the domain decomp tool

einola opened the pull request:
The point is that a user can use the tool to partition a given grid without going to the hassle of installing Zoltan and other prerequisites. It works well enough as is, even if the syntax is not the most obvious (but I added a section in the README.md file).

The only sore point is that I used ``COPY . /decomp`` in the Dockerfile, which triggers a rebuild of the tool every time the Docker image is built after modifying any file in the root directory. The alternatives are to select which files to copy or to pull everything from GitHub. Pulling from GitHub seems a bit much since the source is already there, and copying select files means updating the list if the project changes. So I went for what I considered the least bad solution.

Closes #69

einola committed:
Adds a docker image for the domain decomp tool

The point is that a user can use the tool to partition a given grid
without going to the hassle of installing Zoltan and other
prerequisites. It works well enough as is, even if the syntax is not the
most obvious (but I added a section in the README.md file).

einola committed:
Make the -h message look a bit nicer when using Docker

I copied the `decomp` executable to /usr/bin, so that the
`entrypoint.sh` script can just use `exec decomp`, instead of the full
path. This means that the `Usage` message gives just the name of the
executable, and not the full path to it.

einola commented:
I needed this, so I did it :)

joewallwork reviewed (changes requested):
Nice, thanks for this @einola! I like your use of an entrypoint script. I guess another natural (follow-up) extension to this PR would be to have the nextSIM-DG Docker images(s) pull the `decomp` installation from the one created here.

I hit a build error when I tried to run locally because it was picking up a different `decomp` binary in the `build` directory. `rm -rf build` did the trick and with that the Dockerised tool works nicely.

To avoid others hitting the same issue, perhaps we should put the Dockerfile in a subdirectory (e.g., `Dockerfiles/`), as we did in the nextSIM-DG? Alternatively, the setup could be left as-is, but the gotcha above would be a useful addition to the docs.

einola committed:
Use a temporary directory for building

The executable is copied to /usr/bin anyway, and this avoids accidentally
picking up a build directory from the user (like what happened to Joe).

einola commented on README.md:
````diff
@@ -289,3 +292,33 @@ for each MPI process. The file also defines the variables `X_neighbours(P)`, `X_
+
+### How to run using Docker
+
+First, build the docker image, e.g.
+
+```
+$ docker build . -t decomp
+```
````
Yes, we should build a hierarchy of docker images on nextsimhub. That should be a separate issue, though. Also, I'm using an M1 Mac, which means I can't use x64 docker images like the ones on nextsimhub.

joewallwork replied:
Ah okay, I hadn't thought about that.

einola committed:
Slightly improved README.txt

Tell people why we should use --rm, not just say that it's recommended.

joewallwork reviewed (approved):
Thanks @einola, this should be good to go once you merge in `main` to pick up the Mac CI fix.

einola committed:
Merge remote-tracking branch 'origin/main' into issue69_docker
